lib/pbench/server/api/resources/upload_api.py

```diff
@@ -355,27 +354,95 @@ def put(self, filename: str):
                         f"Failed to write .md5 file '{md5_full_path}'",
                     )
 
-                # Create a file tree object
-                try:
-                    file_tree = FileTree(self.config, self.logger)
-                    self.logger.info(
-                        "Created {} of type {}", file_tree, type(file_tree)
-                    )
-                except Exception:
-                    raise CleanupTime(
-                        HTTPStatus.INTERNAL_SERVER_ERROR, "Unable to map the file tree"
-                    )
+            # Create a file tree object
+            try:
+                file_tree = FileTree(self.config, self.logger)
+            except Exception:
+                raise CleanupTime(
+                    HTTPStatus.INTERNAL_SERVER_ERROR, "Unable to map the file tree"
+                )
 
-                # Move the files to their final location
-                try:
-                    tarball = file_tree.create(controller, tar_full_path)
-                except Exception:
-                    raise CleanupTime(
-                        HTTPStatus.INTERNAL_SERVER_ERROR,
-                        "Unable to create dataset in file system",
+            # Move the files to their final location
+            try:
+                tarball = file_tree.create(controller, tar_full_path)
+            except Exception:
+                raise CleanupTime(
+                    HTTPStatus.INTERNAL_SERVER_ERROR,
+                    f"Unable to create dataset in file system for {tar_full_path}",
+                )
+
+            # From this point, failure will remove the tarball from the file
+            # tree.
+            #
+            # TODO: the Tarball.delete method won't clean up empty controller
+            # directories, but to call FileTree.delete(dataset.name) we'd need
+            # to add parameters to the recovery object.
+            recovery.add(tarball.delete)
+
+            # Now that we have the tarball, extract the dataset timestamp from
+            # the metadata.log file.
+            #
+            # If this fails, the metadata.log is missing or corrupt, so fail
+            # the upload.
+            #
+            # NOTE: we're setting the Dataset "created" timestamp here, but it
+            # won't be committed to the database until the "advance" operation
+            # at the end.
+            #
+            # TODO: Create a Metadata key containing the full metadata.log JSON
+            # returned by `tarball.get_metadata()`; for now we're throwing away
+            # everything except the date, which we need for the main Dataset
+            # object.
```

Comment:
Seems like this should be moved to an issue, and perhaps additional context added to help the reader understand the implications.